### PR TITLE
Remove unnecessary settings argument from SemanticParser

### PR DESCRIPTION
**Pull request by EniddeallA**
This PR fixes issue https://github.com/pyccel/pyccel/issues/1354.

**Commit Summary**

- Remove settings
- Update documentation/doc strings

### CODE REVIEW

File: tests/pyccel/scripts/test2.json [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            "/home/eniddealla/.local/lib/python3.10/site-packages/numpy/core/include"
+        ],
+        "shared_suffix": ".cpython-310-x86_64-linux-gnu.so",
+        "libs": [],
+        "libdirs": [
+            "/usr/lib/python3.10/config-3.10-x86_64-linux-gnu",
+            "/usr/lib/x86_64-linux-gnu"
+        ]
+    }
+}
```
Comment (EmilyBourne) [inline tests/pyccel/scripts/test2.json:66]: Beware, these files are auto-generated by the tests but they should not be committed to the repo. Maybe you have some kind of setting which automatically commits any new file in the folder?

File: pyccel/parser/semantic.py [hunk 2]
```diff
@@ -2709,12 +2696,13 @@ def _visit_Assign(self, expr, **settings):
...
 
-    def _visit_For(self, expr, **settings):
+    def _visit_For(self, expr):
 
         scope = self.create_new_loop_scope()
 
         # treatment of the index/indices
-        iterable = Iterable(self._visit(expr.iterable, **settings))
+        iterable = Iterable(self._visit(expr.iterable))
+        body     = list(expr.body.body)
```
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:2705]: Beware! Bad merge!

**Comment by EmilyBourne** [comment]
See https://numpydoc.readthedocs.io/en/latest/format.html for details of how to format docstrings. Let me know if you need a hand writing any of the docs. I think `_visit` and `annotate` should be ok from the developer docs, but the other functions may be harder if you are not familiar with the code

**Comment by EniddeallA** [comment]
The coverage errors are addressed in the following issues:
#1372
#1390
#1391
#1392
@EmilyBourne What do you think ?

**Comment by EniddeallA** [comment]
The untested codes are:
## `pyccel/parser/semantic.py`
https://github.com/pyccel/pyccel/blob/48df47e8006cb05aa10857325d09b7f6cbaf9dd5/pyccel/parser/semantic.py#L2410-L2417
Coverage related to macro functions. #1372


## `pyccel/parser/semantic.py`
https://github.com/pyccel/pyccel/blob/28aa4a5570ff0948ab354307b062ddc65b9a5118/pyccel/parser/semantic.py#L2444-L2468
Coverage in a CodeBlock in _visit_Assign Function. #1390

## `pyccel/parser/semantic.py`
https://github.com/pyccel/pyccel/blob/28aa4a5570ff0948ab354307b062ddc65b9a5118/pyccel/parser/semantic.py#L2504-L2513
Coverage in FunctionCall in _visit_Assign Function. #1391

## `pyccel/parser/semantic.py`
https://github.com/pyccel/pyccel/blob/28aa4a5570ff0948ab354307b062ddc65b9a5118/pyccel/parser/semantic.py#L2573-L2609
Coverage related to tuples in _visit_Assign Function. #1392

**Review by smhah — APPROVED** [review]
well done

**Review by mkaddani — APPROVED** [review]
well done!

File: pyccel/parser/semantic.py [hunk 3]
```diff
@@ -464,26 +474,35 @@ def _garbage_collector(self, expr):
...
+
+        Parameters
+        ----------
+        expr : data-type
+                Data type.
+
+        Returns
+        -------
+        d_var
+            Dictionary of type inference for expr.
```
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:492]: The type of the parameter is not accurate here, and I don't think "type inference" is very well defined. What about:
```suggestion
        Infer all relevant type information for the expression.

        Create a dictionary describing all the type information that can be
        inferred about the expression `expr`. This includes information about:
        - `datatype`
        - `precision`
        - `rank`
        - `shape`
        - `order`
        - `memory_handling`
        - `cls_base`
        - `is_target`

        Parameters
        ----------
        expr : pyccel.ast.basic.Basic
                An AST object representing an object in the code whose type
                must be determined.

        Returns
        -------
        dict
            Dictionary containing all the type information which was inferred.
```

File: pyccel/parser/semantic.py [hunk 4]
```diff
@@ -1530,14 +1550,25 @@ def _assign_GeneratorComprehension(self, lhs_name, expr, **settings):
...
+        raised.
+        
+        Parameters
+        ----------
+        expr : object
+            Object to visit.
+        
+        Returns
+        -------
+        obj
```
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:1570]: See above. Also fix the argument type. The type should be `pyccel.ast.basic.Basic`

**Review by EmilyBourne — CHANGES_REQUESTED** [review]
Looks good. I just have a few comments about the docstrings

File: pyccel/parser/semantic.py [hunk 5]
```diff
@@ -270,8 +270,18 @@ def program_namespace(self):
...
+        Examine the AST.
+
+        Call to 'self._visit(self.ast)' from the constructor to
+        examine an AST object created by the syntactic stage.
+
+        Returns
+        -------
+        ast
+            Sympy ast.
+        """
```
[the remark this answers is not in the document]
Comment (EniddeallA) [reply pyccel/parser/semantic.py:284]: I found it in the syntactic code while tracing its initial declaration.
https://github.com/pyccel/pyccel/blob/c858fef58aabaff942d9a57a6ad27fc4cd13f399/pyccel/parser/syntactic.py#L152-L152

File: pyccel/parser/semantic.py [hunk 6]
```diff
@@ -1037,40 +1050,44 @@ def _ensure_target(self, rhs, d_lhs):
...
         arr_in_multirets : bool
             If True, rhs has an array in its results, otherwise, it should be set to False.
             It helps when we don't need lhs to be a pointer in case of a returned array in
             a tuple of results.
 
-        settings : dictionary
-            Provided to all _visit_ClassName functions
+        Returns
+        -------
+        lhs
```
[the remark this answers is not in the document]
Comment (EniddeallA) [reply pyccel/parser/semantic.py:1101]: yes, I think I was trying to put both the name and the type.

File: pyccel/parser/semantic.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/parser/semantic.py]: Hmmm, I think I ran into this recently. But I'm not sure I fixed it.

I would suggest this:
```
"""
Convert Python's AST to Pyccel's AST object.

Convert Python's AST to Pyccel's AST object and raise errors
for any unsupported objects.

Results
-------
pyccel.ast.basic.Basic
     The Pyccel AST object.
"""
```
the unused `verbose` argument will need to be removed though

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/parser/semantic.py]: If you don't mind that would be great. Thanks :)

[the remark this answers is not in the document]
Comment (EniddeallA) [reply pyccel/parser/semantic.py]: done

File: pyccel/parser/semantic.py [hunk 7]
```diff
@@ -1407,21 +1424,24 @@ def format_shape(s):
...
+                    The name to which the expression is assigned.
         expr : GeneratorComprehension
+                The GeneratorComprehension node.
 
-        Results
+        Returns
         -------
-        new_expr : CodeBlock
-                   CodeBlock containing the semantic version of the GeneratorComprehension node
+        expr_new
```
[the remark this answers is not in the document]
Comment (EniddeallA) [reply pyccel/parser/semantic.py]: fixed

File: pyccel/parser/semantic.py (hunk 4 above)
Comment (EniddeallA) [reply pyccel/parser/semantic.py]: fixed

File: pyccel/parser/semantic.py [hunk 8]
```diff
@@ -1407,21 +1436,23 @@ def format_shape(s):
...
+                    The name to which the expression is assigned.
         expr : GeneratorComprehension
+                The GeneratorComprehension node.
 
-        Results
+        Returns
         -------
-        new_expr : CodeBlock
-                   CodeBlock containing the semantic version of the GeneratorComprehension node
+        pyccel.ast.variable.Variable
```
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:1455]: Why did you remove the return description?

File: pyccel/parser/semantic.py [hunk 9]
```diff
@@ -1530,14 +1561,24 @@ def _assign_GeneratorComprehension(self, lhs_name, expr, **settings):
...
+        raised.
+        
+        Parameters
+        ----------
+        expr : pyccel.ast.basic.Basic
+            Object to visit.
+        
+        Returns
+        -------
+        pyccel.ast.variable.Variable
```
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:1581]: Missing description of what is returned

Comment (EniddeallA) [reply pyccel/parser/semantic.py:1582]: What do you think of "AST object of type X." as description ?

File: pyccel/parser/semantic.py (hunk 8 above)
Comment (EniddeallA) [reply pyccel/parser/semantic.py:1455]: Wrong input, fixed.

File: pyccel/parser/semantic.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/parser/semantic.py]: You have put the return type as `pyccel.ast.variable.Variable`, but the returns are:
https://github.com/pyccel/pyccel/blob/279c24409e351a3ae8106efc5030a29a96156b57/pyccel/parser/semantic.py#L1551-L1558

These have type `pyccel.ast.functionalexpr.GeneratorComprehension`

File: pyccel/parser/semantic.py (hunk 4 above)
Comment (EmilyBourne) [reply pyccel/parser/semantic.py]: Please add the commit SHA to your message when you fix things so it's easy to check that the changes fix all the comments raised.
Here I don't think the fix is correct. You have put the wrong type

File: pyccel/parser/semantic.py (hunk 8 above)
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1455]: I've just realised that this return type is wrong. See above

File: pyccel/parser/semantic.py (hunk 9 above)
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:1582]: I don't think this is an accurate description. The argument (`expr`) has type `X` but the result sometimes has a different type (e.g. `FunctionCall(PyccelSymbol('shape'), PyccelSymbol('x'))` would return `NumpyShape(Variable('x'))`

File: pyccel/parser/semantic.py [hunk 10]
```diff
@@ -1037,40 +1062,44 @@ def _ensure_target(self, rhs, d_lhs):
...
             If True, rhs has an array in its results, otherwise, it should be set to False.
             It helps when we don't need lhs to be a pointer in case of a returned array in
             a tuple of results.
 
-        settings : dictionary
-            Provided to all _visit_ClassName functions
+        Returns
+        -------
+        pyccel.ast.variable.Variable
+            The representation of the lhs provided by the SyntacticParser.
```
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:1102]: There is an error here:
```suggestion
            The representation of the lhs provided by the SemanticParser.
```

**Review by EmilyBourne — CHANGES_REQUESTED** [review]

File: pyccel/parser/semantic.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (EniddeallA) [reply pyccel/parser/semantic.py]: Fixed in bce4623

File: pyccel/parser/semantic.py (hunk 8 above)
Comment (EniddeallA) [reply pyccel/parser/semantic.py]: It got fixed in [bce4623](https://github.com/pyccel/pyccel/pull/1385/commits/bce46231be73f3c7515d9765ec4808f5e557deda)

File: pyccel/parser/semantic.py (hunk 4 above)
Comment (EniddeallA) [reply pyccel/parser/semantic.py]: the fix was in [279c244](https://github.com/pyccel/pyccel/pull/1385/commits/279c24409e351a3ae8106efc5030a29a96156b57) do you mean the return type "pyccel.ast.variable.Variable" is wrong?

Comment (EmilyBourne) [reply pyccel/parser/semantic.py]: Yes. `_visit` returns objects of lots of different types. E.g. `pyccel.ast.core.Assign` which is not a Variable. My first comment told you the correct type

File: pyccel/parser/semantic.py (hunk 9 above)
Comment (EniddeallA) [reply pyccel/parser/semantic.py:1582]: What would be a better description, should I describe it as just "AST object" ?

File: pyccel/parser/semantic.py (hunk 4 above)
Comment (EniddeallA) [reply pyccel/parser/semantic.py]: I thought you said to fix the argument's (`expr`) type.
Do you mean the return type should be `pyccel.ast.basic.Basic`
Can you please confirm from [279c244](https://github.com/pyccel/pyccel/pull/1385/commits/279c24409e351a3ae8106efc5030a29a96156b57) if the changes are right or wrong?

Comment (EmilyBourne) [reply pyccel/parser/semantic.py]: Ah I understand the confusion now. Both the argument and the return object have the type `pyccel.ast.basic.Basic`. This is the type from which all elements of pyccel's AST inherit. This function takes an AST object and returns an equivalent object with more information about type etc. These two objects are equivalent but do not necessarily have the same type.

In [279c244](https://github.com/pyccel/pyccel/pull/1385/commits/279c24409e351a3ae8106efc5030a29a96156b57) I only see changes to the descriptions. I don't think it is right but this has its own conversation below

File: pyccel/parser/semantic.py [hunk 11]
```diff
@@ -1530,14 +1562,25 @@ def _assign_GeneratorComprehension(self, lhs_name, expr, **settings):
...
+        
+        Parameters
+        ----------
+        expr : pyccel.ast.basic.Basic
+            Object to visit.
+        
+        Returns
+        -------
+        pyccel.ast.variable.Variable
+            AST object of type X.
```
Comment (EmilyBourne) [inline pyccel/parser/semantic.py:1583]:
```suggestion
        expr : pyccel.ast.basic.Basic
            Object to visit of type X.
        
        Returns
        -------
        pyccel.ast.basic.Basic
            AST object which is the semantic equivalent of expr.
```

File: pyccel/parser/semantic.py (hunk 4 above)
Comment (EniddeallA) [reply pyccel/parser/semantic.py]: Fixed in [0d130cb](https://github.com/pyccel/pyccel/pull/1385/commits/0d130cbf9cd3f9e77cb52f2dafeb7c15ae36e28a)

File: CHANGELOG.md [hunk 12]
```diff
@@ -88,5 +88,6 @@ All notable changes to this project will be documented in this file.
 -   #1245 : Reduce false negative test results by using a tolerance to compare floats.
 -   #1272 : Remove use of deprecated NumPy syntax in tests.
 -   #1253 : Provide minimum requirements.
+-   #1385 : Remove settings from `_visit` function.
```
Comment (EmilyBourne) [inline CHANGELOG.md:91]: Could you mark this as `INTERNAL`. It does not affect users:
```suggestion
-   \[INTERNALS\]  #1385 : Remove unused settings keyword arguments from `_visit` function.
```

**Review by EmilyBourne — APPROVED** [review]
Looks good :)

**Review by yguclu — APPROVED** [review]
